File: services/cupom.py
"""
Formata cupons de impressão para pedidos individuais e de convênio.
Saída em texto puro (UTF-8) para impressora térmica de 42 colunas.
"""

from datetime import datetime

LARGURA = 42
SEP_SIMPLES = "─" * LARGURA
SEP_DUPLO   = "═" * LARGURA


def _centralizar(texto: str) -> str:
    return texto.center(LARGURA)


def _linha_kv(chave: str, valor: str) -> str:
    espacos = LARGURA - len(chave) - len(valor)
    if espacos < 1:
        espacos = 1
    return f"{chave}{' ' * espacos}{valor}"


def montar_cupom_individual(pedido: dict, itens: list[dict]) -> str:
    """
    pedido: linha da tabela `pedidos`
    itens:  lista de linhas da tabela `itens_pedido` (geralmente 1 item)
    """
    agora = datetime.now().strftime("%d/%m  %H:%M")
    linhas = [
        _linha_kv(agora, "INDIVIDUAL"),
        SEP_SIMPLES,
    ]

    numero = pedido.get("numero_whatsapp") or ""
    for item in itens:
        nome = (item.get("nome_pessoa") or "CLIENTE").upper()
        mistura = item.get("mistura") or ""
        tamanho = item.get("tamanho") or ""
        linhas.append(f"{nome}  {numero}")
        linhas.append(f"{mistura}  {tamanho}")

        acomps = [a for a in [item.get("acomp_1"), item.get("acomp_2")] if a]
        if acomps:
            linhas.append(" + ".join(a.title() for a in acomps))

        obs = item.get("observacoes") or ""
        if obs.strip():
            linhas.append(f"⚠ {obs}")

    entrega = pedido.get("endereco_entrega") or ""
    hora_ret = pedido.get("hora_retirada") or ""
    if hora_ret:
        linhas.append(f"Retirada ~{hora_ret}")
    elif entrega:
        linhas.append(f"Entrega: {entrega}")

    linhas.append(SEP_SIMPLES)
    return "\n".join(linhas)


def montar_cupom_convenio(pedido: dict, itens: list[dict], nome_empresa: str) -> str:
    """
    pedido:       linha da tabela `pedidos`
    itens:        lista de itens_pedido do lote
    nome_empresa: nome da empresa (de empresas_convenio)
    """
    agora = datetime.now().strftime("%d/%m  %H:%M")
    hora_ent = pedido.get("hora_retirada") or pedido.get("horario_entrega") or "—"

    linhas = [
        _centralizar(f"★ {nome_empresa.upper()} ★"),
        _centralizar(f"Entrega {hora_ent}"),
        SEP_DUPLO,
    ]

    total_valor = 0.0
    for item in itens:
        nome    = (item.get("nome_pessoa") or "").upper()
        mistura = item.get("mistura") or ""
        tamanho = item.get("tamanho") or ""
        valor   = float(item.get("valor_unitario") or 0)
        obs     = item.get("observacoes") or ""

        valor_str = f"R${valor:.2f}".replace(".", ",") if valor else ""

        linhas.append(nome)
        linha_prato = f"  {mistura}  {tamanho}"
        if valor_str:
            linha_prato = _linha_kv(f"  {mistura}  {tamanho}", valor_str)
        linhas.append(linha_prato)

        if obs.strip():
            linhas.append(f"  ⚠ {obs}")

        linhas.append(SEP_SIMPLES)
        total_valor += valor

    forma_pgto = pedido.get("forma_pgto") or "Convênio mensal"
    total_str  = f"R${total_valor:.2f}".replace(".", ",")

    linhas.append(_linha_kv(f"TOTAL: {len(itens)} marmita(s)", total_str))
    linhas.append(f"Forma pgto: {forma_pgto}")
    linhas.append(SEP_DUPLO)

    return "\n".join(linhas)

File: windows_service/GustoImpressao/Cupom.vb
Imports System.Text

Public Module CupomBuilder

    Private Const LARGURA As Integer = 42
    Private ReadOnly SEP_SIMPLES As String = New String("─"c, LARGURA)
    Private ReadOnly SEP_DUPLO As String = New String("═"c, LARGURA)

    Public Function MontarCupomIndividual(pedido As PedidoImpressao) As String
        Dim agora = DateTime.Now.ToString("dd/MM  HH:mm")
        Dim sb As New StringBuilder()

        sb.AppendLine(LinhaKV(agora, "INDIVIDUAL"))
        sb.AppendLine(SEP_SIMPLES)

        Dim fone = FormatarTelefone(pedido.NumeroWhatsapp)

        For Each item In pedido.Itens
            Dim nome = If(String.IsNullOrWhiteSpace(item.NomePessoa), "CLIENTE", item.NomePessoa).ToUpper()
            sb.AppendLine($"{nome}  {fone}")
            sb.AppendLine($"{item.Mistura}  {item.Tamanho}")

            Dim acomps = {item.Acomp1, item.Acomp2}.Where(Function(a) Not String.IsNullOrWhiteSpace(a)).ToArray()
            If acomps.Length > 0 Then
                sb.AppendLine(String.Join(" + ", acomps.Select(Function(a) ToTitleCase(a))))
            End If

            If Not String.IsNullOrWhiteSpace(item.Observacoes) Then
                sb.AppendLine($"⚠ {item.Observacoes}")
            End If
        Next

        If Not String.IsNullOrWhiteSpace(pedido.HoraRetirada) Then
            sb.AppendLine($"Retirada ~{pedido.HoraRetirada}")
        ElseIf Not String.IsNullOrWhiteSpace(pedido.EnderecoEntrega) Then
            sb.AppendLine($"Entrega: {pedido.EnderecoEntrega}")
        End If

        sb.AppendLine(SEP_SIMPLES)
        Return sb.ToString()
    End Function

    Public Function MontarCupomConvenio(pedido As PedidoImpressao, nomeEmpresa As String) As String
        Dim agora = DateTime.Now.ToString("dd/MM  HH:mm")
        Dim horaEnt = If(Not String.IsNullOrWhiteSpace(pedido.HoraRetirada), pedido.HoraRetirada, "—")

        Dim sb As New StringBuilder()
        sb.AppendLine(Centralizar($"★ {nomeEmpresa.ToUpper()} ★"))
        sb.AppendLine(Centralizar($"Entrega {horaEnt}"))
        sb.AppendLine(SEP_DUPLO)

        Dim totalValor As Decimal = 0
        For Each item In pedido.Itens
            Dim nome = If(String.IsNullOrWhiteSpace(item.NomePessoa), "", item.NomePessoa).ToUpper()
            Dim valorStr = If(item.ValorUnitario > 0, $"R${item.ValorUnitario:N2}".Replace(".", ","), "")

            sb.AppendLine(nome)
            Dim linhaPrato = $"  {item.Mistura}  {item.Tamanho}"
            sb.AppendLine(If(String.IsNullOrEmpty(valorStr), linhaPrato, LinhaKV(linhaPrato, valorStr)))

            Dim acomps = {item.Acomp1, item.Acomp2}.Where(Function(a) Not String.IsNullOrWhiteSpace(a)).ToArray()
            If acomps.Length > 0 Then
                sb.AppendLine("  " & String.Join(" + ", acomps.Select(Function(a) ToTitleCase(a))))
            End If

            If Not String.IsNullOrWhiteSpace(item.Observacoes) Then
                sb.AppendLine($"  ⚠ {item.Observacoes}")
            End If

            sb.AppendLine(SEP_SIMPLES)
            totalValor += item.ValorUnitario
        Next

        Dim totalStr = $"R${totalValor:N2}".Replace(".", ",")
        sb.AppendLine($"TOTAL: {pedido.Itens.Count} marmita(s)  {totalStr}")
        sb.AppendLine($"Forma pgto: {If(String.IsNullOrWhiteSpace(pedido.FormaPgto), "Convênio mensal", pedido.FormaPgto)}")
        sb.AppendLine(SEP_DUPLO)
        Return sb.ToString()
    End Function

    Private Function LinhaKV(chave As String, valor As String) As String
        Dim espacos = LARGURA - chave.Length - valor.Length
        If espacos < 1 Then espacos = 1
        Return chave & New String(" "c, espacos) & valor
    End Function

    Private Function Centralizar(texto As String) As String
        Return texto.PadLeft((LARGURA + texto.Length) \ 2).PadRight(LARGURA)
    End Function

    Private Function FormatarTelefone(numero As String) As String
        If String.IsNullOrEmpty(numero) Then Return ""
        If numero.StartsWith("55") AndAlso numero.Length >= 12 Then
            numero = numero.Substring(2)
        End If
        If numero.Length = 11 Then
            Return $"({numero.Substring(0, 2)}) {numero(2)}.{numero.Substring(3, 4)}-{numero.Substring(7)}"
        End If
        Return numero
    End Function

    Private Function ToTitleCase(s As String) As String
        If String.IsNullOrEmpty(s) Then Return s
        Return Char.ToUpper(s(0)) & s.Substring(1).ToLower()
    End Function

End Module
